cpp/src/reductions/reduction.cuh

```diff
@@ -0,0 +1,126 @@
+
+  using ScalarType = cudf::experimental::scalar_type_t<OutputType>;
+  auto s = new ScalarType(std::move(dev_result), true, stream, mr); //only for string_view, data is copied
+  return std::unique_ptr<scalar>(s);
+}
+
+/** --------------------------------------------------------------------------*
+ * @brief compute reduction by the compound operator (reduce and transform)
+ *
+ * @param[in] d_in      the begin iterator
+ * @param[in] num_items the number of items
+ * @param[in] op        the reduction operator 
+ * @param[in] valid_count   the intermediate operator argument 1
+ * @param[in] ddof      the intermediate operator argument 2
+ * @param[in] stream    cuda stream
+ * @returns   scalar    output scalar in device memory
+ *
+ * The reduction operator must have intermediate::compute_result() method.
+ * This method performs reduction using binary operator `Op::Op` and transforms the
+ * result to `OutputType` using `compute_result()` transform method.
+ *
+ * @tparam Op               the reduction operator
+ * @tparam InputIterator    the input column iterator
+ * @tparam OutputType       the output type of reduction
+ * ----------------------------------------------------------------------------**/
+template <typename Op, typename InputIterator, typename OutputType, 
+  typename IntermediateType=typename thrust::iterator_value<InputIterator>::type>
+  //FIXME: adding this causes reduction.cuh:118:401:   error : ‘__T290’ was not declared in this scope
+  //typename std::enable_if_t< std::is_base_of<cudf::experimental::reduction::op::CompoundOp<Op>, Op>::value, int> = 0>
+std::unique_ptr<scalar> reduce(InputIterator d_in, cudf::size_type num_items, Op op, 
+  cudf::size_type valid_count,
+  cudf::size_type ddof,
+  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  typename Op::Op binary_op{};
+  IntermediateType identity = Op::Op::template identity<IntermediateType>();
+  rmm::device_scalar<IntermediateType> intermediate_result{identity, stream, mr};
+
+  // Allocate temporary storage
+  rmm::device_buffer d_temp_storage;
+  size_t temp_storage_bytes = 0;
+  cub::DeviceReduce::Reduce(d_temp_storage.data(), temp_storage_bytes, d_in, intermediate_result.data(), num_items, binary_op, identity, stream);
+  d_temp_storage = rmm::device_buffer{temp_storage_bytes, stream, mr};
+
+  // Run reduction
+  cub::DeviceReduce::Reduce(d_temp_storage.data(), temp_storage_bytes, d_in, intermediate_result.data(), num_items, binary_op, identity, stream);
+
+  // compute the result value from intermediate value in device
+  using ScalarType = cudf::experimental::scalar_type_t<OutputType>;
+  auto result = new ScalarType(OutputType{0}, true, stream, mr);
+  thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+   intermediate_result.data(), 1,
+   [dres=result->data(), valid_count, ddof] __device__ (auto i)
+   { *dres = Op::template compute_result<OutputType>(i, valid_count, ddof); } );
```

Comment:
My feeling is that copying device -> host and doing the finalization step on the host would be faster. Are there reduction benchmarks you can try and see?

Reply:
I don't have reduction benchmarks. 
Here again it needs to copy the data to device to return `unique_ptr<scalar>`. That's why I implemented like this. 
Old code did what you suggest. @harrism suggested in https://github.com/rapidsai/cudf/pull/3273#discussion_r347168758 to avoid copying back and forth. I implemented a single thread thrust call.